compiler-core/src/docs/printer.rs

```diff
@@ -0,0 +1,578 @@
+        break_("(", "(")
+            .append(join(arguments, break_(",", ", ")))
+            .nest_if_broken(INDENT)
+            .append(break_(",", ""))
+            .append(")")
+    }
+
+    fn type_arguments<'a>(arguments: impl IntoIterator<Item = Document<'a>>) -> Document<'a> {
+        break_("", "")
+            .append(join(arguments, break_(",", ", ")))
+            .nest_if_broken(INDENT)
+            .append(break_(",", ""))
+            .group()
+            .surround("(", ")")
+    }
+
+    fn type_(&mut self, type_: &Type) -> Document<'static> {
+        match type_ {
+            Type::Named {
+                package,
+                module,
+                name,
+                args,
+                ..
+            } => {
+                let name = self.named_type_name(package, module, name);
+                if args.is_empty() {
+                    name
+                } else {
+                    name.append(Self::type_arguments(
+                        args.iter().map(|argument| self.type_(argument)),
+                    ))
+                }
+            }
+            Type::Fn { args, return_ } => docvec![
+                self.keyword("fn"),
+                Self::type_arguments(args.iter().map(|argument| self.type_(argument))),
+                " -> ",
+                self.type_(return_)
+            ],
+            Type::Tuple { elements } => docvec![
+                "#",
+                Self::type_arguments(elements.iter().map(|element| self.type_(element))),
+            ],
+            Type::Var { type_ } => match type_.as_ref().borrow().deref() {
+                TypeVar::Link { type_ } => self.type_(type_),
+
+                TypeVar::Unbound { id } | TypeVar::Generic { id } => {
+                    let name = self.type_variable(*id);
+                    self.variable(name)
+                }
+            },
+        }
+    }
+
+    fn type_variable(&mut self, id: u64) -> EcoString {
+        if let Some(name) = self.names.get_type_variable(id) {
+            return name.clone();
+        }
+
+        if let Some(name) = self.printed_type_variables.get(&id) {
+            return name.clone();
+        }
+
+        loop {
+            let name = self.next_letter();
+            if !self.printed_type_variable_names.contains(&name) {
+                _ = self.printed_type_variable_names.insert(name.clone());
+                _ = self.printed_type_variables.insert(id, name.clone());
+                return name;
+            }
+        }
+    }
+
+    // Copied from the `next_letter` method of the `type_::printer`.
+    fn next_letter(&mut self) -> EcoString {
+        let alphabet_length = 26;
+        let char_offset = b'a';
+        let mut chars = vec![];
+        let mut n;
+        let mut rest = self.next_type_variable_id;
+
+        loop {
+            n = rest % alphabet_length;
+            rest /= alphabet_length;
```

Comment:
Don't use these "mutate this variable with this operation" operators please. Prefer the regular assignment syntax as it is clearer.